### PR TITLE
KVStore general tests

### PR DESCRIPTION
**Pull request by theamirocohen**
[ ] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [x] Test update
    [ ] Breaking change

### CODE REVIEW

**Comment by theamirocohen** [comment]
@davidsaada @yossi2le @offirko @dannybenor 
initial tests, please review

File: features/storage/TESTS/kvstore/general_tests/main.cpp [hunk 1]
```diff
@@ -0,0 +1,961 @@
...
+static void set_key_null()
+{
+    int res = kvstore->set(NULL, data, data_size, 0);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : key length over key max size
+static void set_key_length_exceeds_max()
+{
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = "key_name";
```
Comment (davidsaada) [inline features/storage/TESTS/kvstore/general_tests/main.cpp]: This should not error. As key is a null terminated string, buffer size has no relevance. You will need to set a large string to make it error.

File: features/storage/TESTS/kvstore/general_tests/main.cpp [hunk 2]
```diff
@@ -0,0 +1,961 @@
...
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+
+    res = kvstore->set_finalize(handle);
+    TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
+}
+
+
+/*----------------setup------------------*/
+
+Case cases[] = {
```
Comment (davidsaada) [inline features/storage/TESTS/kvstore/general_tests/main.cpp:896]: It is better to define a failure handler that will make the tests continue. You have a lot of tests here, and it's better to have all of them run, even if one ore more fail.

File: features/storage/TESTS/kvstore/general_tests/main.cpp [hunk 3]
```diff
@@ -0,0 +1,961 @@
...
+    for (int i = 0; i < num_of_threads; i++) {
+        threadStatus = kvstore_thread[i].join();
+        if (threadStatus != 0) {
+            utest_printf("\nthread %d join failed!", i + 1);
+        }
+    }
+    
+    for (int i = 0; i < num_of_threads; i++) {
+        int res = kvstore->get(keys[i], buffer, buffer_size, &actual_size, 0);
+        TEST_ASSERT_EQUAL(MBED_SUCCESS, res);
```
Comment (davidsaada) [inline features/storage/TESTS/kvstore/general_tests/main.cpp:137]: Why not check the returned data here as well?

File: features/storage/TESTS/kvstore/general_tests/main.cpp [hunk 4]
```diff
@@ -0,0 +1,961 @@
...
+static void get_key_null()
+{
+    int res = kvstore->get(NULL, buffer, buffer_size, &actual_size, 0);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : key length over key max size
+static void get_key_length_exceeds_max()
+{
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = "key_name";
```
Comment (davidsaada) [inline features/storage/TESTS/kvstore/general_tests/main.cpp]: Same here: Key doesn't exceed max size.

**Review by davidsaada — CHANGES_REQUESTED** [review]
Made a partial review. Will continue tomorrow.

File: features/storage/TESTS/kvstore/general_tests/main.cpp [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+    char key_max[KVStore::MAX_KEY_SIZE + 1] = {0};
+    memset(key_max, '*', KVStore::MAX_KEY_SIZE);
+    int res = kvstore->get_info(key_max, &info);
+    TEST_ASSERT_EQUAL(MBED_ERROR_INVALID_ARGUMENT, res);
+}
+
+//bad params : &info is null
+static void get_info_info_null()
+{
+    int res = kvstore->get_info(key, NULL);
```
Comment (davidsaada) [inline features/storage/TESTS/kvstore/general_tests/main.cpp:457]: This should actually work, though not defined explicitly. We can use this function to check whether a key exists, therefore the info argument should defaults to NULL. Will change the base class also.

@offirko Please change this accordingly also in FileSystemStore.

Comment (offirko) [reply features/storage/TESTS/kvstore/general_tests/main.cpp:457]: Updated my code and tests

**Comment by davidsaada** [comment]
Went through all tests. Very good coverage.
Fixed all TDBStore errors failing tests (please uncomment them and remove failure comments).
All that's left to fix is all the "exceed max key size" tests and the null info tests.

**Pull request closed without merging**